In [0]:
#Problem 1

In [0]:
!pip install sqlalchemy
!pip install psycopg2
!pip install psycopg2-binary

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import seaborn as sns
import math
import statsmodels.api as sm
from scipy import stats
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from scipy.stats.mstats import winsorize
from sqlalchemy import create_engine
import warnings



# import warnings
warnings.filterwarnings('ignore')

In [0]:
# Database credentials
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

# use the credentials to start a connection
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df= pd.read_sql_query('select * from weatherinszeged', con=engine)

# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

In [0]:
#explore the dataset
df

date  ...                            dailysummary
0     2006-03-31 22:00:00+00:00  ...       Partly cloudy throughout the day.
1     2006-03-31 23:00:00+00:00  ...       Partly cloudy throughout the day.
2     2006-04-01 00:00:00+00:00  ...       Partly cloudy throughout the day.
3     2006-04-01 01:00:00+00:00  ...       Partly cloudy throughout the day.
4     2006-04-01 02:00:00+00:00  ...       Partly cloudy throughout the day.
...                         ...  ...                                     ...
96448 2016-09-09 17:00:00+00:00  ...  Partly cloudy starting in the morning.
96449 2016-09-09 18:00:00+00:00  ...  Partly cloudy starting in the morning.
96450 2016-09-09 19:00:00+00:00  ...  Partly cloudy starting in the morning.
96451 2016-09-09 20:00:00+00:00  ...  Partly cloudy starting in the morning.
96452 2016-09-09 21:00:00+00:00  ...  Partly cloudy starting in the morning.

[96453 rows x 12 columns]

In [0]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
date                   96453 non-null datetime64[ns, UTC]
summary                96453 non-null object
preciptype             96453 non-null object
temperature            96453 non-null float64
apparenttemperature    96453 non-null float64
humidity               96453 non-null float64
windspeed              96453 non-null float64
windbearing            96453 non-null float64
visibility             96453 non-null float64
loudcover              96453 non-null float64
pressure               96453 non-null float64
dailysummary           96453 non-null object
dtypes: datetime64[ns, UTC](1), float64(8), object(3)
memory usage: 8.8+ MB


In [0]:
#checking for null values
df.isnull().sum()

date                   0
summary                0
preciptype             0
temperature            0
apparenttemperature    0
humidity               0
windspeed              0
windbearing            0
visibility             0
loudcover              0
pressure               0
dailysummary           0
dtype: int64

In [0]:
#describe the data
df.describe()


temperature  apparenttemperature  ...  loudcover      pressure
count  96453.000000         96453.000000  ...    96453.0  96453.000000
mean      11.932678            10.855029  ...        0.0   1003.235956
std        9.551546            10.696847  ...        0.0    116.969906
min      -21.822222           -27.716667  ...        0.0      0.000000
25%        4.688889             2.311111  ...        0.0   1011.900000
50%       12.000000            12.000000  ...        0.0   1016.450000
75%       18.838889            18.838889  ...        0.0   1021.090000
max       39.905556            39.344444  ...        0.0   1046.380000

[8 rows x 8 columns]

In [0]:
df=df.assign(apparent_and_temp=df["apparenttemperature"]-df["temperature"])

In [0]:
df.head()

date  ... apparent_and_temp
0 2006-03-31 22:00:00+00:00  ...         -2.083333
1 2006-03-31 23:00:00+00:00  ...         -2.127778
2 2006-04-01 00:00:00+00:00  ...          0.000000
3 2006-04-01 01:00:00+00:00  ...         -2.344444
4 2006-04-01 02:00:00+00:00  ...         -1.777778

[5 rows x 13 columns]

In [0]:
# Y is the target variable
Y = df['apparent_and_temp']

# X is the feature set
X = df[['humidity','windspeed']]

# We add a constant to the model as it's a best practice
# to do so every time!
X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:      apparent_and_temp   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Fri, 20 Mar 2020   Prob (F-statistic):               0.00
Time:                        04:08:37   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.0

 
 r**2 =0.288 and adjusted r**2  is 0.228  They are not satisfacory as the values are too low. This means that the model is not explaining 
 much info about the outcome.

In [0]:
# including interaction of humidity and windspeed
df["humidity_windspeed"]=df.humidity*df.windspeed
# Y is the target variable
Y = df['apparent_and_temp']

# X is the feature set
X = df[['humidity','windspeed', "humidity_windspeed"]]

# We add a constant to the model as it's a best practice
# to do so every time!
X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:      apparent_and_temp   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Fri, 20 Mar 2020   Prob (F-statistic):               0.00
Time:                        04:08:37   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                  0.0839      0

R-squared is 0.341  Adj. R-squared is now 0.341. yes the model imporoves upon the previous one.

In [0]:
# add visibility
# Y is the target variable
Y = df['apparent_and_temp']

# X is the feature set
X = df[['humidity','windspeed', "visibility"]]

# We add a constant to the model as it's a best practice
# to do so every time!
X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:      apparent_and_temp   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Fri, 20 Mar 2020   Prob (F-statistic):               0.00
Time:                        04:08:38   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.0

Analysis: R-squared:0.304 Adj R-squared: 0.303 . Both r squared and adjusted r squared increased. rsquared is higher with the interaction term added compared to the visibility term. Thus the interaction tem contribures more than having the visibility.


For AIC and BIC the lower values are better.
The firsst model AIC:3.409e+05BIC:3.409e+05 Second 
The second model model AIC:3.334e+05 BIC: 3.334e+05 Best 
The third model AIC:3.388e+05BIC: 3.388e+05 Third
Clearly adding the interaciton term made a big difference.

In [0]:
# Problem 2
# Database credentials
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

# use the credentials to start a connection
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df= pd.read_sql_query('select * from houseprices', con=engine)

# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

In [0]:
df.columns

Index(['id', 'mssubclass', 'mszoning', 'lotfrontage', 'lotarea', 'street',
       'alley', 'lotshape', 'landcontour', 'utilities', 'lotconfig',
       'landslope', 'neighborhood', 'condition1', 'condition2', 'bldgtype',
       'housestyle', 'overallqual', 'overallcond', 'yearbuilt', 'yearremodadd',
       'roofstyle', 'roofmatl', 'exterior1st', 'exterior2nd', 'masvnrtype',
       'masvnrarea', 'exterqual', 'extercond', 'foundation', 'bsmtqual',
       'bsmtcond', 'bsmtexposure', 'bsmtfintype1', 'bsmtfinsf1',
       'bsmtfintype2', 'bsmtfinsf2', 'bsmtunfsf', 'totalbsmtsf', 'heating',
       'heatingqc', 'centralair', 'electrical', 'firstflrsf', 'secondflrsf',
       'lowqualfinsf', 'grlivarea', 'bsmtfullbath', 'bsmthalfbath', 'fullbath',
       'halfbath', 'bedroomabvgr', 'kitchenabvgr', 'kitchenqual',
       'totrmsabvgrd', 'functional', 'fireplaces', 'fireplacequ', 'garagetype',
       'garageyrblt', 'garagefinish', 'garagecars', 'garagearea', 'garagequal',
       'garagecond', 'paved

In [0]:
df=pd.concat([df,pd.get_dummies(df.mszoning,prefix="mszoning",drop_first=True)], axis=1)
df=pd.concat([df,pd.get_dummies(df.street,prefix="street",drop_first=True)], axis=1)
dummy_column_names=list(pd.get_dummies(df.mszoning,prefix="mszoning",drop_first=True).columns)
dummy_column_names=dummy_column_names + list(pd.get_dummies(df.street,prefix="street",drop_first=True).columns)

In [0]:
X=df[["overallqual", "grlivarea" ,"garagecars","garagearea", "totalbsmtsf"] + dummy_column_names]
Y=df["saleprice"]
# We add a constant to the model as it's a best practice
# to do so every time!
X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:              saleprice   R-squared:                       0.769
Model:                            OLS   Adj. R-squared:                  0.767
Method:                 Least Squares   F-statistic:                     482.0
Date:                Fri, 20 Mar 2020   Prob (F-statistic):               0.00
Time:                        04:08:38   Log-Likelihood:                -17475.
No. Observations:                1460   AIC:                         3.497e+04
Df Residuals:                    1449   BIC:                         3.503e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.173e+05    1.8e+04     -6.502      

F-test is F-statistic is 482.0 This does not make much sense unless we are comparing with something else. R-squared:0.769Adj. R-squared:0.767 77% of the variance in thetarget variable is explained by the model. 23% isn't. There is room for improvement. 
AIC: 3.497e+04 BIC: 3.503e+04 these cannot be assesed without something to compare with. 

In [0]:
df["total_sf"]=df['firstflrsf'] + df['secondflrsf'] + df['totalbsmtsf']

df['int_over_sf']=df['total_sf'] *df['overallqual']


X = df[['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'total_sf', 'int_over_sf'] + dummy_column_names]
Y=np.log1p(df["saleprice"])
# We add a constant to the model as it's a best practice
# to do so every time!
X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:              saleprice   R-squared:                       0.832
Model:                            OLS   Adj. R-squared:                  0.831
Method:                 Least Squares   F-statistic:                     651.8
Date:                Fri, 20 Mar 2020   Prob (F-statistic):               0.00
Time:                        04:10:56   Log-Likelihood:                 570.75
No. Observations:                1460   AIC:                            -1117.
Df Residuals:                    1448   BIC:                            -1054.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const           9.8514      0.089    110.698      

In [0]:
# we see general improvement in the second case